grafana/grafana
# Prometheus: Fix interpolation of $__rate_interval variable

**What this PR does / why we need it**:

This PR tackles 3 things:
1. It fixes https://github.com/grafana/grafana/issues/43720 that was happening due to using `intervalv2.FormatDuration` that rounds the duration. So if the duration was `5m30s` it became `5m`.
2. This is probably an edge case, but it fixes the calculation of  `$__rate_interval` when  `$__rate_interval` is used both. in `interval` and in `expr`. If it is used in both, we need to just use the `$__rate_interval` and not calculate it again.
3. Fixes using of `$__rate_interval` in interval.

I have decided to create helper functions and move different logics into different functions to make it more clear what is happening.
**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/43720

**How to test/reproduce:**
To reproduce/test first issue, you can set min interval 5m and then the rate interval should be 5m + scrape interval (in my case I have 30 sec scrape interval). Therefore correct interpolated rate interval should be 5m30s, while in. the past it would be evaluated to 5m:

https://user-images.githubusercontent.com/30407135/149511333-8d3391ec-a403-456b-b822-373e8c3ea0ba.mov

2. When you use rate interval in interval, it should be interpolated to 4x scrape interval (in my case 30 sec) => 2 min

https://user-images.githubusercontent.com/30407135/149512008-406a366e-555a-4e7a-a345-d86a714afddc.mov
